src/basic/namespace-util.c

```diff
@@ -210,14 +213,65 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+         * process. We can use that for file system mounts and similar.
+         * If the owner of the new namespace is specified, setuid to it in the child. Use a pipe to sync and
+         * ensure setuid() happens first, then unshare(), then the mapping update. */
 
-        r = safe_fork("(sd-mkuserns)", FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_NEW_USERNS, &pid);
+        if (pipe2(pipe_fd, 0) < 0)
+                return log_error_errno(errno, "Failed to open bidirectional pipe: %m");
+
+        /* If we have to set a different owner we have to setuid before unsharing the new namespace, otherwise
+         * the kernel will refuse the operation, so don't fork directly in the new user namespace in that case.
+         * If we don't have to change the owner then it easier and we can let the clone set up the namespace
+         * instead via the apposite flag. */
+        if (owner_uid == UID_INVALID)
+                flags |= FORK_NEW_USERNS;
+
+        r = safe_fork_full("(sd-mkuserns)", (int[]) { pipe_fd[1] }, 1, flags, &pid);
         if (r < 0)
                 return r;
-        if (r == 0)
-                /* Child. We do nothing here, just freeze until somebody kills us. */
+        if (r == 0) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
+                if (owner_uid != UID_INVALID) {
+                        if (owner_gid != GID_INVALID) {
+                                r = RET_NERRNO(setresgid(owner_gid, owner_gid, owner_gid));
+                                if (r < 0)
+                                        goto child_fail;
+                        }
+
+                        r = RET_NERRNO(setresuid(owner_uid, owner_uid, owner_uid));
+                        if (r < 0)
+                                goto child_fail;
+
+                        r = RET_NERRNO(unshare(CLONE_NEWUSER));
+                        if (r < 0)
+                                goto child_fail;
+                }
+
+                pipe_fd[1] = safe_close(pipe_fd[1]);
                 freeze();
+                _exit(EXIT_SUCCESS);
+
+        child_fail:
+                (void) write(pipe_fd[1], &r, sizeof(r));
+                pipe_fd[1] = safe_close(pipe_fd[1]);
+                _exit(EXIT_FAILURE);
+        }
+
+        pipe_fd[1] = safe_close(pipe_fd[1]);
+
+        if (owner_uid != UID_INVALID) {
+                r = fd_wait_for_event(pipe_fd[0], POLLHUP|POLLIN, USEC_PER_SEC * 10);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        return -ETIMEDOUT;
+                if (r & POLLIN) {
+                        if (read(pipe_fd[0], &r, sizeof(r)) == sizeof(r))
+                                return log_error_errno(r, "Failed to create user namespace: %m");
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to create user namespace.");
+                }
+        }
```

Comment:
hmm, i'd really send a `r = 0` integer from the child to the parent on success. Because otherwise, if the child dies abnormally it will look like everything is in order since we'l get POLLHUP then just like if everything was OK.

hence: the child should always send a message up, r == 0 in case of success, and something else on failure. and parent should always read the thing, and if it can't consider that an error